### PR TITLE
[feature][expressions] Add "length3D" function to return the 3D length of a LineGeometry type geometry

### LINKED ISSUE #41115: lineStringZ length calculation function
In the QGIS field calculator there are over 100 geometric functions but none to calculate the length of `LineStringZ`.
- `$lenght` calculates the length respecting the project settings (ellipsoidal length)
- `lenght ($geometry)` calculates the 2D Cartesian length

it would be very useful to be able to calculate the 3D Cartesian length as visible from the information through icon 

image

### PR DESCRIPTION
**Pull request by agiudiceandrea**
Calculates the cartesian 3D length of a geometry line object. If the geometry is not a 3D line object, it returns its 2D length.

The function uses QgsLineString::length3D() to perfom the calculation.

Fixes https://github.com/qgis/QGIS/issues/41115

### CODE REVIEW

File: tests/src/core/testqgsexpression.cpp [hunk 1]
```diff
@@ -941,6 +941,10 @@ class TestQgsExpression: public QObject
       QTest::newRow( "length line" ) << "length(geom_from_wkt('LINESTRING(0 0, 4 0)'))" << false << QVariant( 4.0 );
       QTest::newRow( "length polygon" ) << "length(geom_from_wkt('POLYGON((0 0, 4 0, 4 2, 0 2, 0 0))'))" << false << QVariant();
       QTest::newRow( "length point" ) << "length(geom_from_wkt('POINT(0 0)'))" << false << QVariant();
+      QTest::newRow( "length3D lineZ" ) << "length3D(geom_from_wkt('LINESTRINGZ(0 0 0, 3 0 4)'))" << false << QVariant( 5.0 );
+      QTest::newRow( "length3D line" ) << "length3D(geom_from_wkt('LINESTRING(0 0, 4 0)'))" << false << QVariant( 4.0 );
+      QTest::newRow( "length3D polygon" ) << "length3D(geom_from_wkt('POLYGON((0 0, 4 0, 4 2, 0 2, 0 0))'))" << false << QVariant();
+      QTest::newRow( "length3D point" ) << "length3D(geom_from_wkt('POINT(0 0)'))" << false << QVariant();
```
Comment (lbartoletti) [inline tests/src/core/testqgsexpression.cpp:947]: Can you add also the resulsts of:

```
length3D(geom_from_wkt('POLYGON((0 0, 4 0, 4 2, 0 2, 0 0))'))
length3D(geom_from_wkt('LINESTRINGZ EMPTY'))
```

Comment (agiudiceandrea) [reply tests/src/core/testqgsexpression.cpp:947]: Hi @lbartoletti, `length3D(geom_from_wkt('POLYGON((0 0, 4 0, 4 2, 0 2, 0 0))'))` is already tested and correctly returns QVariant().
It seems the other geometry functions, like length, perimeter, area, are not tested against EMPTY wkt geometries. However `length3D(geom_from_wkt('LINESTRINGZ EMPTY'))` returns 0 like `length(geom_from_wkt('LINESTRING EMPTY'))`.

File: src/core/expression/qgsexpressionfunction.cpp [hunk 2]
```diff
@@ -1356,6 +1356,32 @@ static QVariant fcnLength( const QVariantList &values, const QgsExpressionContex
   return QVariant( str.length() );
 }
 
+static QVariant fcnLength3D( const QVariantList &values, const QgsExpressionContext *, QgsExpression *parent, const QgsExpressionNodeFunction * )
+{
+  QgsGeometry geom = QgsExpressionUtils::getGeometry( values.at( 0 ), parent );
+
+  if ( geom.type() != QgsWkbTypes::LineGeometry )
```
Comment (nyalldawson) [inline src/core/expression/qgsexpressionfunction.cpp:1363]: Curved geometries will also return a line geometry type, so you'll need to handle these somehow (eg by segmentizing and then calculating the distance. Or by working out the math for true 3d distance for curves! 😁)

Comment (agiudiceandrea) [reply src/core/expression/qgsexpressionfunction.cpp:1363]: Hi @nyalldawson , thanks for the review and for pointing that out. Could it be a correct approach to always cast to QgsCurve and then use curveToLine() to extract a QgsLineString representation? I don't know the "curve" world well.
Anyway If you have a better idea on how to implement this function, we could disregard this PR.

Comment (roya0045) [reply src/core/expression/qgsexpressionfunction.cpp:1363]: I suggest looking into https://github.com/qgis/QGIS/blob/16493ca562d4961471ec5d7af087e6541b88e46d/src/core/geometry/qgsgeometryutils.cpp#L755 you could either alter it to add z/m values or use the base for the lenght and afterward add surface distance. You can see how it is used in QgsCircularString.

I think that would be adequate.

Comment (agiudiceandrea) [reply src/core/expression/qgsexpressionfunction.cpp:1363]: @roya0045 thanks for the good suggestion!

Anyway I think "length3D" should behave like the "length" and "perimeter" (and "area") functions.

It seems they don't use `QgsCircularString::length()` to calculate the length of a CircularString, so the calculated value is equal to the value of `geometry.constGet().curveToLine().length()`.

If fact, with a R=100 circonference/circle, in the field calculator:
- `length(geom_from_wkt('CompoundCurve (CircularString (100 200, 200 100, 100 0, 0 100, 100 200))'))` and
- `perimeter(geom_from_wkt('CurvePolygon (CompoundCurve (CircularString (100 200, 200 100, 100 0, 0 100, 100 200)))'))` 

both return 628.3105558829233 (< 2\*π\*R)

like
```python
circonference=QgsGeometry.fromWkt('CompoundCurve (CircularString (100 200, 200 100, 100 0, 0 100, 100 200))')
circonference.constGet().curveToLine().length()
```
or
```python
circonference.length()
```

while
```python
circonference.constGet().length()
```
returns 628.3185307179587 (=2\*π\*r).

Comment (agiudiceandrea) [reply src/core/expression/qgsexpressionfunction.cpp:1363]: @nyalldawson, now length3D function handles curved geometries too.

Comment (roya0045) [reply src/core/expression/qgsexpressionfunction.cpp:1363]: Nice work! Though I'm unsure what make the same tests keep failling in Travis and if something is now broken (unrelated to this PR)

**Comment by nyalldawson** [comment]
Superseded by https://github.com/qgis/QGIS/pull/41780

**Pull request closed without merging**